[WFLY-12758]:  Integrating OpenTracing

ehsavoie opened the pull request:
* Configuring jaeger tracers.
* Need to obtain the configuration service in the TraceInitializer.
* opentracing-interceptors dependency is set to provided.

Jira: https://issues.jboss.org/browse/WFLY-12758
      https://issues.jboss.org/browse/WFLY-12812

ehsavoie commented:
@jmesnil @pferraro  @zhantemirov Could you please review this one ?

ehsavoie commented:
Galleon failure is due to the fact that jaeger is not provisioned if no tracer is defined in the subsystem which is now required with

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/JaegerTracerConfigurationAddHandler.java:
```diff
@@ -0,0 +1,66 @@
+        if (outboundSocketBindingName != null) {
+            outboundSocketBindingSupplier = builder.requires(OUTBOUND_SOCKET_BINDING_CAPABILITY.getCapabilityServiceName(outboundSocketBindingName));
+        } else {
+            outboundSocketBindingSupplier = new Supplier<OutboundSocketBinding>() {
```
You can replace this anonymous class with:
org.wildfly.common.function.Functions.constantSupplier(null);

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemDefinition.java:
```diff
@@ -32,42 +32,77 @@
+            .of(DEFAULT_TRACER_CAPABILITY_NAME, true)
+            .build();
+
+    public static final RuntimeCapability<Void> TRACER_CAPABILITY = RuntimeCapability.Builder
```
This duplicates JaegerTracerConfigurationDefinition.TRACER_CAPABILITY.  You can delete this.

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemDefinition.java:
```diff
@@ -32,42 +32,77 @@
                 .setAddHandler(SubsystemAdd.INSTANCE)
-                .setRemoveHandler(ReloadRequiredRemoveStepHandler.INSTANCE)
+                .setRemoveHandler(new SubsystemRemoveHandler())
                 .setCapabilities(OPENTRACING_CAPABILITY)
```
Should you not also register DEFAULT_TRACER_CAPABILITY?

ehsavoie replied:
No because this default is only set if the default attribute (which is nullable) is set

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemDefinition.java:
```diff
@@ -32,42 +32,77 @@
             .build();
 
+    public static final RuntimeCapability<Void> DEFAULT_TRACER_CAPABILITY = RuntimeCapability.Builder
+            .of(DEFAULT_TRACER_CAPABILITY_NAME, true)
```
This should probably define a service type of TracerConfiguration.  It should not be dynamic.

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemDefinition.java:
```diff
@@ -32,42 +32,77 @@
+
+    public static final SimpleAttributeDefinition DEFAULT_TRACER = SimpleAttributeDefinitionBuilder
+            .create(TracerConfigurationConstants.DEFAULT_TRACER, ModelType.STRING, true)
+            .setCapabilityReference(TRACER_CAPABILITY_NAME)
```
After you additionally register DEFAULT_TRACER_CAPABILITY (see below), you won't be able to use this method.  You'll instead need to use:
setCapabilityReference(TRACER_CAPABILITY_NAME, DEFAULT_TRACER_CAPABILITY)

ehsavoie replied:
This would be a Resource capability instead of one bounded to the value of an attribute

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemDefinition.java:
```diff
@@ -79,4 +114,39 @@ public void registerAdditionalRuntimePackages(final ManagementResourceRegistrati
+        @Override
+        protected void recordCapabilitiesAndRequirements(OperationContext context, AttributeDefinition attributeDefinition,
+                ModelNode newValue, ModelNode oldValue) {
+            super.recordCapabilitiesAndRequirements(context, attributeDefinition, newValue, oldValue);
```
I don't think you want to call the super implementation, as you are handling capability registration conditionally.

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemDefinition.java:
```diff
@@ -79,4 +114,39 @@ public void registerAdditionalRuntimePackages(final ManagementResourceRegistrati
+
+        @Override
+        protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource) throws OperationFailedException {
+            super.recordCapabilitiesAndRequirements(context, operation, resource);
```
I don't think you want to call the super implementation, as you are handling capability registration conditionally.

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemAdd.java:
```diff
@@ -19,41 +19,60 @@
+
+    @Override
+    protected void recordCapabilitiesAndRequirements(OperationContext context, ModelNode operation, Resource resource) throws OperationFailedException {
+        super.recordCapabilitiesAndRequirements(context, operation, resource);
```
I don't think you want to call the super implementation, as you are handling capability registration conditionally.

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemAdd.java:
```diff
@@ -19,41 +19,60 @@
+        super.recordCapabilitiesAndRequirements(context, operation, resource);
+        ModelNode defaultTracer = DEFAULT_TRACER.resolveModelAttribute(context, operation);
+        if (defaultTracer.isDefined()) {
+            context.registerCapability(RuntimeCapability.Builder.of(DEFAULT_TRACER_CAPABILITY.getDynamicName(defaultTracer.asString()), false).build());
```
DEFAULT_TRACER_CAPABILITY shouldn't be dynamic.

ehsavoie replied:
I need it to be able to have the tracer name in the deployment

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemDefinition.java:
```diff
@@ -79,4 +114,39 @@ public void registerAdditionalRuntimePackages(final ManagementResourceRegistrati
+                ModelNode newValue, ModelNode oldValue) {
+            super.recordCapabilitiesAndRequirements(context, attributeDefinition, newValue, oldValue);
+            if (oldValue.isDefined()) {
+                context.deregisterCapability(DEFAULT_TRACER_CAPABILITY.getDynamicName(oldValue.asString()));
```
DEFAULT_TRACER_CAPABILITY shouldn't be dynamic.

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemDefinition.java:
```diff
@@ -79,4 +114,39 @@ public void registerAdditionalRuntimePackages(final ManagementResourceRegistrati
+                context.deregisterCapability(DEFAULT_TRACER_CAPABILITY.getDynamicName(oldValue.asString()));
+            }
+            if (newValue.isDefined()) {
+                context.registerCapability(RuntimeCapability.Builder.of(DEFAULT_TRACER_CAPABILITY.getDynamicName(newValue.asString()), false).build());
```
DEFAULT_TRACER_CAPABILITY shouldn't be dynamic.

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemDefinition.java:
```diff
@@ -79,4 +114,39 @@ public void registerAdditionalRuntimePackages(final ManagementResourceRegistrati
+            super.recordCapabilitiesAndRequirements(context, operation, resource);
+            ModelNode defaultTracer = DEFAULT_TRACER.resolveModelAttribute(context, resource.getModel());
+            if (defaultTracer.isDefined()) {
+                context.deregisterCapability(RuntimeCapability.Builder.of(DEFAULT_TRACER_CAPABILITY.getDynamicName(defaultTracer.asString()), false).build().getName());
```
DEFAULT_TRACER_CAPABILITY shouldn't be dynamic.

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/SubsystemExtension.java:
```diff
@@ -42,10 +43,17 @@
-        return getResourceDescriptionResolver(true, keyPrefix);
+        String[] prefixes = new String[keyPrefix.length + 1];
+        prefixes[0] = SUBSYSTEM_NAME;
+        System.arraycopy(keyPrefix, 0, prefixes, 1, keyPrefix.length);
```
This seems a bit unnecessary.  Any way to avoid it?

ehsavoie replied:
Looking into it but this is not relevant to this PR :)

ehsavoie replied:
no part of this PR ;)  but fixing that

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/TracerAttributes.java:
```diff
@@ -0,0 +1,110 @@
+            .setRequired(false)
+            .setAllowedValues("JAEGER", "B3")
+            .setAttributeGroup("codec-configuration")
+            .setRestartAllServices()
```
Why must this require a restart of all services?  Same question for all attributes below.

ehsavoie replied:
Because this is a configuration of a tracer, not a tracer itself thus you need to recreate a Tracer instance to update it which means restarting the deployments that have the tracer already injected.

pferraro replied:
But we would only need to restart those deployments with a dependency on the specific tracer configuration, not all deployments, right?
In that case you can use setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES).

ehsavoie replied:
ok

bstansberry replied:
AttributeAccess.Flag.RESTART_RESOURCE_SERVICES is something of an anti-pattern as a typical user won't understand that bouncing a service means bouncing the deployment. And bouncing a deployment isn't graceful.

This seems like a possible future enhancement.

ehsavoie replied:
ok so keeping my restartAllServices

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/TracerAttributes.java:
```diff
@@ -0,0 +1,110 @@
+            .build();
+
+    public static final SimpleAttributeDefinition SAMPLER_TYPE = SimpleAttributeDefinitionBuilder.create(TracerConfigurationConstants.SAMPLER_TYPE, ModelType.STRING, true)
+            .setAllowedValues("const", "probabilistic", "ratelimiting", "remote")
```
Is there an enum for these values?

ehsavoie replied:
Yes in jaeger's code but I don't want to depends on it

pferraro replied:
So, in that case, you'll probably want to create your own Enum.

ehsavoie replied:
I'd rather not as it doesn't add anything since the value is passed to jaeger directly

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/TracingDeploymentProcessor.java:
```diff
@@ -66,8 +78,41 @@ public void deploy(DeploymentPhaseContext deploymentPhaseContext) throws Deploym
+        Object resolvedTracer = null;
+        try {
+            final Module module = deploymentUnit.getAttachment(Attachments.MODULE);
+            if (module != null) {
```
When is this ever null?

ehsavoie replied:
I don't know :) Maybe I can remove this

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/TracingDeploymentProcessor.java:
```diff
@@ -110,10 +168,30 @@ private JBossWebMetaData getJBossWebMetaData(DeploymentUnit deploymentUnit) {
+            return null;
+        }
+        for (ParamValueMetaData param : jbossWebMetaData.getContextParams()) {
+            if (TracerInitializer.SMALLRYE_OPENTRACING_TRACER_CONFIGURATION.equals(param.getParamName())) {
```
Having not read the opentracing spec thoroughly...
Is this really how users should define opentracing configuration?  I would expect this configuration to be defined via MP Config. Configuration via servlet context parameters is very retro.

ehsavoie replied:
This is how this was done in the initial integration. Tracer configuration is not really specified in opentracing specs.

pferraro replied:
To align with vendor-specific configuration for other MP specs, I would suggest using vendor specific MP config namespace. e.g. "mp.opentracing.extensions.tracer.configuration" (or similar).

ehsavoie replied:
Except for those that were already defined

pferraro reviewed (changes requested):
Cursory review.  More to come.

bstansberry commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/JaegerTracerConfigurationAddHandler.java:
```diff
@@ -0,0 +1,66 @@
+    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+        super.performRuntime(context, operation, model);
+        ServiceName serviceName = TRACER_CAPABILITY.getCapabilityServiceName(context.getCurrentAddressValue());
+        ServiceBuilder builder = context.getServiceTarget().addService(serviceName);
```
This should should use context.getCapabilityServiceTarget.

ehsavoie replied:
ok

bstansberry commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/JaegerTracerConfigurationAddHandler.java:
```diff
@@ -0,0 +1,66 @@
+        String outboundSocketBindingName = TracerAttributes.SENDER_BINDING.resolveModelAttribute(context, model).asStringOrNull();
+        Supplier<OutboundSocketBinding> outboundSocketBindingSupplier;
+        if (outboundSocketBindingName != null) {
+            outboundSocketBindingSupplier = builder.requires(OUTBOUND_SOCKET_BINDING_CAPABILITY.getCapabilityServiceName(outboundSocketBindingName));
```
Using CapabilityServiceBuilder from CapabilityServiceTarget removes the need for a fake OUTBOUND_SOCKET_BINDING_CAPABILITY just to generate a ServiceName.

ehsavoie replied:
ok

bstansberry commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/microprofile/opentracing/smallrye/WildFlyTracerFactory.java:
```diff
@@ -0,0 +1,110 @@
+ */
+public class WildFlyTracerFactory {
+
+    private static final Map<ServiceName, Configuration> CONFIGURATIONS = new HashMap<>();
```
This can be Map<String, Configuration> with null used for the compatibility tracer. That eliminates all the conversion from simple tracer name string -> RuntimeCapability -> ServiceName.

bstansberry commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/OpentracingConfigurationService.java:
```diff
@@ -0,0 +1,62 @@
+    @SuppressWarnings("unchecked")
+    public static final void installTracerConfigurationService(ServiceBuilder builder, TracerConfiguration config, ServiceName serviceName) {
+        builder.setInstance(new OpentracingConfigurationService(WildFlyTracerFactory.registerTracer(serviceName), config));
+        builder.install();
```
If WildFlyTracerFactory.registerTracer took a simple config name string and not a ServiceName, then the OCS constructor is
```
public OpentracingConfigurationService(TracerConfiguration config, String configName) {
   this.config = config;
   this.consumer = WildFlyTracerFactory.registerTracer(configName)
}
```
and then this method can go away with the instance construction and the install moved to the code that is doing the rest of the manipulation of the ServiceBuilder.

If a getName() method was valid for the TracerConfiguration interface, then you just need 
```
public OpentracingConfigurationService(TracerConfiguration config) { 
```

ehsavoie replied:
ok and I can even remove that Service class

bstansberry commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/JaegerTracerConfigurationDefinition.java:
```diff
@@ -0,0 +1,78 @@
+public class JaegerTracerConfigurationDefinition extends SimpleResourceDefinition {
+
+    public static final RuntimeCapability<Void> TRACER_CAPABILITY = RuntimeCapability.Builder
+            .of(TRACER_CAPABILITY_NAME, true, OpentracingConfigurationService.class)
```
+1.

ehsavoie replied:
ok

bstansberry commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/microprofile/opentracing-smallrye/main/module.xml:
```diff
@@ -39,6 +40,8 @@
         <module name="javax.enterprise.api" />
+        <module name="io.opentracing.opentracing-api"/>
         <module name="org.wildfly.microprofile.opentracing-smallrye"/>
+        <module name="io.jaegertracing.jaeger" optional="true"/>
```
Please add a comment noting what makes it required, e.g.

<!-- Only needed if a jaeger-tracer is configured -->

ehsavoie replied:
Well it will be required as it is required by adding the legacy behaviour now.

bstansberry commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/microprofile/opentracing-smallrye/main/module.xml:
```diff
@@ -25,6 +25,11 @@
 
     <dependencies>
+        <module name="org.jboss.logging"/>
+        <module name="org.jboss.as.controller"/>
```
See my comment about WildFlyTraceFactory.CONFIGURATIONS using simple String keys, which would remove the need for RuntimeCapability in this module, which removes the need for this dependency.

You need to add org.jboss.dmr though.

bstansberry commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/microprofile/opentracing-smallrye/main/module.xml:
```diff
@@ -25,6 +25,11 @@
     <dependencies>
+        <module name="org.jboss.logging"/>
+        <module name="org.jboss.as.controller"/>
+        <module name="org.jboss.as.server"/>
```
I don't see the use of this.

bstansberry commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/microprofile/opentracing-smallrye/main/module.xml:
```diff
@@ -25,6 +25,11 @@
+        <module name="org.jboss.logging"/>
+        <module name="org.jboss.as.controller"/>
+        <module name="org.jboss.as.server"/>
+        <module name="org.wildfly.security.elytron"/>
```
I don't see the use of this.

ehsavoie commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/OpentracingConfigurationService.java:
```diff
@@ -0,0 +1,62 @@
+ * Service providing the Tracer configuration.
+ * @author Emmanuel Hugonnet (c) 2019 Red Hat, Inc.
+ */
+public class OpentracingConfigurationService implements Service {
```
I agree

zhantemirov commented:
From QE point of view tests are okay, thank you. Additional test cases will be added by me later.

ehsavoie commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/microprofile/opentracing/smallrye/TracerInitializer.java:
```diff
@@ -42,6 +40,7 @@
 public class TracerInitializer implements ServletContextListener {
```
I agree, I have something working for this

bstansberry commented on microprofile/opentracing-extension/src/main/resources/org/wildfly/extension/microprofile/opentracing/LocalDescriptions.properties:
```diff
@@ -1,3 +1,33 @@
+microprofile-opentracing-smallrye.tracer.read-timeout=Tracer read timeout.
+microprofile-opentracing-smallrye.tracer.sampler=Tracer sampler rate.
+microprofile-opentracing-smallrye.tracer.propagation=List of formats to use for propagating the trace context.
+microprofile-opentracing-smallrye.tracer.sampler-param=The Jaeger sampler parameter.
```
This needs a more complete explanation.

bstansberry commented on microprofile/opentracing-extension/src/main/resources/org/wildfly/extension/microprofile/opentracing/LocalDescriptions.properties:
```diff
@@ -1,3 +1,33 @@
+microprofile-opentracing-smallrye.tracer.message-max-bytes=Tracer message max size.
+microprofile-opentracing-smallrye.tracer.max-request=Tracer max request.
+microprofile-opentracing-smallrye.tracer.read-timeout=Tracer read timeout.
+microprofile-opentracing-smallrye.tracer.sampler=Tracer sampler rate.
```
rate?

kabir commented:
Retest this please

ehsavoie commented on feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/microprofile/opentracing-smallrye/main/module.xml:
```diff
@@ -25,6 +25,9 @@
 
     <dependencies>
+        <module name="org.jboss.logging"/>
+        <module name="org.jboss.dmr"/>
```
ok

ehsavoie commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/TracingDeploymentDefinition.java:
```diff
@@ -0,0 +1,53 @@
+
+    @Override
+    public void registerAttributes(ManagementResourceRegistration resourceRegistration) {
+        super.registerAttributes(resourceRegistration);
```
I need the name of the referenced tracer in the deployment processor

pferraro commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/JaegerTracerConfigurationAddHandler.java:
```diff
@@ -0,0 +1,63 @@
+        }
+        TracerConfiguration config = new JaegerTracerConfiguration(context, operation, outboundSocketBindingSupplier);
+        builder.setInstance(Service.newInstance(WildFlyTracerFactory.registerTracer(config.getName()), config));
+        builder.install();
```
Services should be installed ON_DEMAND whenever possible.  Is there specific reason for this to be ACTIVE?  It seems to me that it should only start when a deployment needs it.

zhantemirov commented:
Additional test cases have been added to the PR and combined with existing ones. 
From QE point of view the test side looks good, thanks.

darranl commented:
Just adding the missing_reqs label as community docs need to be ready at the same time as the feature.

darranl commented:
@ehsavoie Did your latest push address everything in the review from @pferraro - if so I can dismiss the review.

ehsavoie commented:
@darranl everything except the parser

ehsavoie committed:
[WFLY-12758]:  Integrating OpenTracing

* Configuring jaeger tracers.
* Need to obtain the configuration service in the TraceInitializer.
* opentracing-interceptors dependency is set to provided.
* no more memory leak
* Adding some documentation.

Jira: https://issues.jboss.org/browse/WFLY-12758
      https://issues.jboss.org/browse/WFLY-12812
      https://issues.redhat.com/browse/WFLY-12486

zhantemirov commented:
@ehsavoie I can't see tests presented in https://github.com/ehsavoie/wildfly/pull/7/ PR.

ehsavoie commented:
@zhantemirov sorry, fixed

jamezp commented:
I've merged #12878 which requires a rebase of this PR.

ehsavoie committed:
[WFLY-12758]: Integrating OpenTracing

* Adding OpenTracing subsystem integration tests

bstansberry commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/microprofile/opentracing/smallrye/TracerConfigurationConstants.java:
```diff
@@ -0,0 +1,52 @@
+    public static final String SAMPLER_MANAGER_HOST_PORT = "sampler-manager-host-port";
+
+    public static final String SENDER_AGENT_BINDING = "sender-binding";
+    public static final String SENDER_AGENT_HOST = "sender-host";
```
This is no longer used.

bstansberry commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/microprofile/opentracing/smallrye/TracerConfigurationConstants.java:
```diff
@@ -0,0 +1,52 @@
+
+    public static final String SENDER_AGENT_BINDING = "sender-binding";
+    public static final String SENDER_AGENT_HOST = "sender-host";
+    public static final String SENDER_AGENT_PORT = "sender-port";
```
This is no longer used.

bstansberry reviewed (changes requested):
Hi @ehsavoie -- last round of change requests, I hope. ;)

If these could be in by my Monday morning that would be great as I'd like to get it merged as early as possible Monday.

ehsavoie commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/TracingDependencyProcessor.java:
```diff
@@ -30,17 +32,64 @@
+                if (param.getParamValue() != null && !param.getParamValue().startsWith(TRACER_CAPABILITY_NAME)) {
+                    return TRACER_CAPABILITY.getCapabilityServiceName(param.getParamValue());
+                }
+                return ServiceName.parse(param.getParamValue());
```
I agree

ehsavoie commented on microprofile/opentracing-smallrye/src/main/java/org/wildfly/microprofile/opentracing/smallrye/WildFlyTracerFactory.java:
```diff
@@ -0,0 +1,119 @@
+        return null;
+    }
+
+    public static void registerCompatibilityTracer(TracerConfiguration config) {
```
ok

ehsavoie commented on microprofile/opentracing-extension/src/main/java/org/wildfly/extension/microprofile/opentracing/TracingDeploymentProcessor.java:
```diff
@@ -137,10 +125,51 @@ private String getServiceName(DeploymentUnit deploymentUnit) {
+                if (param.getParamValue() != null && !param.getParamValue().startsWith(TRACER_CAPABILITY_NAME)) {
+                    return TRACER_CAPABILITY.getDynamicName(param.getParamValue());
+                }
+                return param.getParamValue();
```
ok
